### PR TITLE
Add options for replicated weights reducer and refine tests

### PR DESCRIPTION
**Pull request by 0xWJ**
This PR introduces an opt-in path to generate weight reducers for replicated weights, carrying an nreplicas value through IR → codegen → runtime so gradients can be divided after reduction. It also expands the test suite to validate the new flag’s effects on generated code and training.
The main goal of this feature is to make sure gradients cross ranks are exactly the same after each step.

### CODE REVIEW

File: nnscaler/runtime/adapter/reducer.py [hunk 1]
```diff
@@ -762,6 +777,11 @@ def __init__(self, ranks: List[int],
         self._zero_use_reduce_scatter = zero_use_reduce_scatter
         self._zero_param_level_sharding = zero_param_level_sharding and self._zero > 0
         self._align_size: int = align_size
+        self._nreplicas: int = nreplicas
+
+        if not isinstance(self._nreplicas, int) or self._nreplicas < 1:
+            raise ValueError(f"nreplicas should be an integer greater than or equal to 1, but got {self._nreplicas}")
+
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply nnscaler/runtime/adapter/reducer.py:787]: fixed

File: tests/parallel_module/test_gencode_reducer.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        r"nnscaler.runtime.adapter.nn.identity_allreduce",
+        instance_name = 'noreplicated_weights',
+    )
+    # generated reducer:
+    # self.wreducer80 = nnscaler.runtime.adapter.Reducer(ranks=[0, 2], reduce_op='sum', async_op=async_op, zero=0, max_bucket_size_bytes=max_bucket_size_bytes, zero_use_reduce_scatter=zero_use_reduce_scatter, zero_param_level_sharding=zero_param_level_sharding,zero_ngroups=1,nreplicas=1)
+
+    m = NormLinearModule()
+    m.train()
+    parallelize(
+        m,
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/parallel_module/test_gencode_reducer.py]: fixed

File: tests/parallel_module/test_gencode_reducer.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        gen_savedir=tmp_path,
+        load_module=False,
+        reuse='override',
+        instance_name = 'replicated_weights',
+    )
+
+    # weight is partitioned, so the reducer should have rank=[0, 2] and nreplicas=1
+    assert _gencode_contains(tmp_path, NormalAddMulMultirefModule, 0,
+        r"nnscaler.runtime.adapter.Reducer\(ranks=\[0, 1, 2, 3\].*nreplicas=2\)",
+        instance_name = 'replicated_weights',
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/parallel_module/test_gencode_reducer.py]: fixed.

File: tests/runtime/test_reducer.py [hunk 4]
```diff
@@ -165,6 +165,25 @@ def test_reducer_build():
     assert buckets[4]._aligned_numel == 4
 
 
+@mock_reducer_env(0, 2)
+def test_reducer_nreplicas():
+    """Test that nreplicas is correctly passed to buckets and applied to gradients."""
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/runtime/test_reducer.py]: fixed

File: tests/parallel_module/test_gencode_reducer.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    # all rank reducer, with nreplicas=2
+    assert _gencode_contains(tmp_path, NormLinearModule, 0,
+        r"nnscaler.runtime.adapter.Reducer\(ranks=\[0, 1, 2, 3\].*nreplicas=2\)",
+        instance_name = 'replicated_weights',
+    )
+    # will have identity allreduce for replicated weights
+    assert _gencode_contains(tmp_path, NormLinearModule, 0,
+        r"nnscaler.runtime.adapter.nn.identity_allreduce",
+        instance_name = 'replicated_weights',
+    )
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/parallel_module/test_gencode_reducer.py:177]: fixed.

File: tests/graph/gener/test_reducer_gen.py [hunk 6]
```diff
@@ -80,11 +81,12 @@ def test_cross_segment_weight_reducer():
...
+    assert reducers[0].nreplicas == 1
 
 
 @pytest.mark.skipif(not torch.cuda.is_available(), reason='lack of gpu devices')
 def test_replicate_shared_param():
-
+    # default: reducer_replicated_weights=False, no reducers for replicated weights
     graph = build_graph()
     for node in graph.select(ntype=IRFwOperation):
         sn1, sn2 = graph.replicate(node, 2)
```
[the remark this answers is not in the document]
Comment (0xWJ) [reply tests/graph/gener/test_reducer_gen.py:93]: fixed

File: nnscaler/runtime/adapter/reducer.py [hunk 7]
```diff
@@ -740,6 +754,10 @@ def __init__(self, ranks: List[int],
             zero_param_level_sharding (bool): whether to use parameter-level sharding in ZeRO
                 This flag is required when use parameter-level optimizers(like Muon)
             align_size (int): the alignment size in bytes for each parameter
+            nreplicas (int): divisor applied to gradients after all-reduce.
+                For replicated weights where each rank already has the full gradient,
+                set this to the number of ranks so the summed gradient is averaged.
+                Default is 1 (no division).
```
Comment (lynex) [inline nnscaler/runtime/adapter/reducer.py:760]: My understanding: for a replicated weight, replicas within a scale unit (same micro-batch) produce identical gradients, while different scale units (different data shards) produce gradients that must be summed. So nreplicas is set to plan_ngpus to cancel only the intra-scale-unit duplication, while preserving the cross-scale-unit sum. Is this correct?

Without this flag (default), replica consistency is maintained by identity_allreduce (identity in forward, all-reduce in backward), and the full cross-rank reducer is skipped to save communication. Is that right?

So the new flag intentionally adds a redundant all-reduce (over all ranks) purely to guarantee bitwise-identical gradients across ranks (determinism), trading communication cost for exact consistency — rather than fixing a correctness bug. Is that the intended motivation? Shall we file an issue for future optimization that avoid redudant communication by carefully planing the reducing orders, etc.

Given the default path already avoids this extra communication, in which scenarios is enabling this flag actually recommended? (e.g. debugging divergence, strict reproducibility, specific optimizers?)

File: tests/parallel_module/test_gencode.py [hunk 8]
```diff
@@ -383,11 +383,11 @@ def forward(self, x, attr):
         return x + getattr(attr, 'a')
 
 
-def print_gencode(cubesave_dir, module_class, index=0):
+def print_gencode(cubesave_dir, module_class, index=0, instance_name=None):
```
Comment (lynex) [inline tests/parallel_module/test_gencode.py:386]: Where is the new arg instance_name been used?

File: tests/parallel_module/test_gencode_reducer.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
+import torch
+import torch.nn.functional as F
+
+
+import nnscaler
+from nnscaler.parallel import parallelize, ComputeConfig
+from tests.utils import replace_all_device_with
+
+from .test_gencode import _gencode_contains, print_gencode
```
Comment (lynex) [inline tests/parallel_module/test_gencode_reducer.py:9]: print_gencode seems not used

File: tests/parallel_module/test_gencode.py (hunk 8 above)
Comment (0xWJ) [reply tests/parallel_module/test_gencode.py:386]: Line 390

File: tests/parallel_module/test_gencode_reducer.py (hunk 9 above)
Comment (0xWJ) [reply tests/parallel_module/test_gencode_reducer.py:9]: Yes. it is only used in debug

File: tests/parallel_module/test_gencode.py (hunk 8 above)
Comment (lynex) [reply tests/parallel_module/test_gencode.py:386]: I mean anywhere that specify instance_name when calling print_gencode?

File: nnscaler/runtime/adapter/reducer.py (hunk 7 above)
Comment (0xWJ) [reply nnscaler/runtime/adapter/reducer.py:760]: My understanding: for a replicated weight, replicas within a scale unit (same micro-batch) produce identical gradients, while different scale units (different data shards) produce gradients that must be summed. So nreplicas is set to plan_ngpus to cancel only the intra-scale-unit duplication, while preserving the cross-scale-unit sum. Is this correct?
-- Yes, but not always set it to `plan_ngpus` , especially when PP is used.

Without this flag (default), replica consistency is maintained by identity_allreduce (identity in forward, all-reduce in backward), and the full cross-rank reducer is skipped to save communication. Is that right?
-- Not always. `identity_allreduce` is inserted for intermediate grad.
for example, 4 gpus , plan_ngpus=2, if the weight is replicated.
x = op(a, weight)
the reducer will be on rank (0, 2) and (1, 3)
So the grad on 0/1 maybe different. Here we will generate reducer on rank (0, 1, 2, 3) with replicas = 2

`identity_allreduce` is a little complex, and this PR doesn't fix it.



So the new flag intentionally adds a redundant all-reduce (over all ranks) purely to guarantee bitwise-identical gradients across ranks (determinism), trading communication cost for exact consistency — rather than fixing a correctness bug. Is that the intended motivation? Shall we file an issue for future optimization that avoid redudant communication by carefully planing the reducing orders, etc.

-- That's right. Please file an issue if you want. But as I said before, this PR is not for `identity_allreduce` case. And You can't fix the problem by carefully planing the reducing order, because the inconsistency can also be introduced by non-deterministic ops (like linear).

Given the default path already avoids this extra communication, in which scenarios is enabling this flag actually recommended? (e.g. debugging divergence, strict reproducibility, specific optimizers?)

-- I think we should recommend this flag for most scenarios.

File: tests/parallel_module/test_gencode.py (hunk 8 above)
Comment (0xWJ) [reply tests/parallel_module/test_gencode.py:386]: no. I only use it when I debug the test case (to view the generated code in debug window)

Comment (lynex) [reply tests/parallel_module/test_gencode.py:386]: better remove such dead code later

Comment (0xWJ) [reply tests/parallel_module/test_gencode.py:386]: No, I need it. it is very useful in debuging.

**Review by lynex — APPROVED** [review]
